{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c3448ed9-ec9d-4662-92f2-49e00b10e6b2",
   "metadata": {},
   "source": [
    "### Custom mirge3 Library\n",
    "mirge3 library include 2/6 nucleotides to the 5'/3' ends of each mature sequence. This script...\n",
    "1. modifies that library by adding a 3rd nucleotide to the 5'-end. \n",
    "2. Ensures the added nucleotides match the assembly.\n",
    "3. Ensures that the positions of the miRNA are reasonable (e.g. validates position against miRBase).\n",
    "4. Uses gnomAD to get SNPs (e.g. does not use mirge3 SNPs).\n",
    "5. Rebuilds index.Libs (e.g. ebwt) files. \n",
    "6. Rebuilds merges file (for SNPs).\n",
    "\n",
    "**Before Running**:  \n",
    "- Go here (https://www.ncbi.nlm.nih.gov/assembly/GCA_000001405.15) and download Refseq assembly.\n",
    "- gunzip -d fna.gz\n",
    "- samtools faidx GRCh38.p13_genomic.fna\n",
    "- Create fasta file from index.lib  \n",
    "`bowtie-inspect miRge3_Lib/human/index.Libs/human_mirna_miRBase > miRBase_annotations.fasta`\n",
    "\n",
    "**Steps**: \n",
    "1. Grab annotations from mirge3 library (seperate SNPs).\n",
    "2. Update annotations and check sequence against assembly.\n",
    "3. Validate annotations against miRBase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3dfd13cf-d601-4cee-8e94-1370f0d97338",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import subprocess as SP\n",
    "import re\n",
    "\n",
    "samtools = Path(r\"/storage/enc/vladlab/project_MDD/bin/samtools-1.10/samtools\")\n",
    "GRCh38 = Path(r\"GCA_000001405.15_GRCh38_genomic.fna\")\n",
    "\n",
    "mirge3_anno = \"/storage/enc/vladlab/project_MDD/scripts_main/23_0103_get_miRNA_annotations/miRBase_annotations.fasta\"\n",
    "miRBase = \"/storage/enc/vladlab/project_MDD/bin/miRge3_Lib/human_23_0528/annotation.Libs/human_miRBase.gff3\"\n",
    "                             \n",
    "\n",
    "chromosome_to_GenBank = {\"1\":\"CM000663.2\", \n",
    "                       \"2\":\"CM000664.2\",\n",
    "                       \"3\":\"CM000665.2\",\n",
    "                       \"4\":\"CM000666.2\",\n",
    "                       \"5\":\"CM000667.2\",\n",
    "                       \"6\":\"CM000668.2\",\n",
    "                       \"7\":\"CM000669.2\",\n",
    "                       \"8\":\"CM000670.2\",\n",
    "                       \"9\":\"CM000671.2\",\n",
    "                       \"10\":\"CM000672.2\",\n",
    "                       \"11\":\"CM000673.2\",\n",
    "                       \"12\":\"CM000674.2\",\n",
    "                       \"13\":\"CM000675.2\",\n",
    "                       \"14\":\"CM000676.2\",\n",
    "                       \"15\":\"CM000677.2\",\n",
    "                       \"16\":\"CM000678.2\",\n",
    "                       \"17\":\"CM000679.2\",\n",
    "                       \"18\":\"CM000680.2\",\n",
    "                       \"19\":\"CM000681.2\",\n",
    "                       \"20\":\"CM000682.2\",\n",
    "                       \"21\":\"CM000683.2\",\n",
    "                       \"22\":\"CM000684.2\",\n",
    "                       \"X\":\"CM000685.2\", \n",
    "                       \"Y\":\"CM000686.2\",\n",
    "                        \"KI270908.1\":\"KI270908.1\",\n",
    "                        \"KV766192.1\":\"KV766192.1\"}\n",
    "\n",
    "\n",
    "# Output\n",
    "output = \"miRBase_annotations.fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7edeebef-6456-4274-9697-b9928e58d090",
   "metadata": {},
   "source": [
    "### 1. Grab annotations from mirge3 library (seperate SNPs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c0574edc-5964-4333-b100-da18c0e60160",
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_annotations_from_miRge3(fasta_file):\n",
    "    anno_dict = {} # key: name,   value: (chromosome, start, stop, strand, sequence, 5'-end, 3'-end)\n",
    "    \n",
    "    with open(fasta_file, 'r') as f_in:\n",
    "        to_record = False\n",
    "        name = None\n",
    "        chromsoome = None\n",
    "        start = None\n",
    "        stop = None\n",
    "        strand = None\n",
    "        org_sequence = None\n",
    "        prune_sequence = None\n",
    "        end5 = None\n",
    "        end3 = None\n",
    "        \n",
    "        \n",
    "        for line in f_in:\n",
    "            line_split = line.split()\n",
    "            if to_record and line.startswith(\">\"): \n",
    "                end5 = org_sequence[0:2]\n",
    "                end3 = org_sequence[-6:]\n",
    "                anno_dict[name] = [chromosome, start, stop, strand, prune_sequence, end5, end3]\n",
    "                to_record = False\n",
    "            \n",
    "            if line.startswith(\">\"):\n",
    "                to_record = True \n",
    "                name = line_split[0][1::]\n",
    "                \n",
    "                # Grab chromosome\n",
    "                if \"chr\" in line_split[1]: # Note: Have unlocalized regions, like KI270908.1\n",
    "                    chromosome = line_split[1][3::]\n",
    "                else:\n",
    "                    chromosome = line_split[1] \n",
    "                positions = line_split[3].split(\":\")\n",
    "                strand = positions[1]\n",
    "                start_stop = positions[2].split('-')\n",
    "                if strand == \"+\":\n",
    "                    start = int(start_stop[0]) + 2\n",
    "                    stop = int(start_stop[1]) - 6\n",
    "                elif strand == \"-\":\n",
    "                    start = int(start_stop[0]) + 6\n",
    "                    stop = int(start_stop[1]) - 2\n",
    "                else:\n",
    "                    print(f\"ERROR!\")\n",
    "            else: \n",
    "                org_sequence = line.strip()\n",
    "                prune_sequence = org_sequence[2:-6]\n",
    "                \n",
    "        # Record last entry\n",
    "        if to_record:\n",
    "            end5 = org_sequence[0:2]\n",
    "            end3 = org_sequence[-6:]\n",
    "            anno_dict[name] = [chromosome, start, stop, strand, prune_sequence, end5, end3]\n",
    "        \n",
    "                    \n",
    "            \n",
    "    return anno_dict\n",
    "\n",
    "mirge3_anno_dict = {} # key: name,   value: (chromosome, start, stop, strand, sequence, end5, end3)\n",
    "mirge3_anno_dict = grab_annotations_from_miRge3(mirge3_anno)\n",
    "\n",
    "# Fixes \n",
    "mirge3_anno_dict[\"hsa-miR-3714-5p\"][0] = \"3\"               # Position on KV766192.1, which isn't part of GCA_000001405.15\n",
    "mirge3_anno_dict[\"hsa-miR-3714-5p\"][1] = 16933195\n",
    "mirge3_anno_dict[\"hsa-miR-3714-5p\"][2] = 16933217\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-1231-5p.SNPA\"][0] = \"1\"          # Other SNP correct\n",
    " \n",
    "mirge3_anno_dict[\"hsa-miR-3198-3p.SNPC\"][0] = \"22\"         # Other SNP correct\n",
    "mirge3_anno_dict[\"hsa-miR-3198-3p.SNPC\"][1] = 17764190\n",
    "mirge3_anno_dict[\"hsa-miR-3198-3p.SNPC\"][2] = 17764211\n",
    "mirge3_anno_dict[\"hsa-miR-3198-3p.SNPC\"][3] = \"-\"\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-1302-3p\"][2] = 30458             # Stop position wrong\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-1303-3p\"][0] = \"5\"               # All wrong\n",
    "mirge3_anno_dict[\"hsa-miR-1303-3p\"][1] =  154685826\n",
    "mirge3_anno_dict[\"hsa-miR-1303-3p\"][2] =  154685848\n",
    "mirge3_anno_dict[\"hsa-miR-1303-3p\"][3] = \"+\"  \n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-1256-5p\"][1] = 43237316          # Start position wrong\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-548u-3p\"][0] = \"6\"               # All wrong\n",
    "mirge3_anno_dict[\"hsa-miR-548u-3p\"][1] =  57390179\n",
    "mirge3_anno_dict[\"hsa-miR-548u-3p\"][2] =  57390202 \n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-378g\"][1] =  26696355            # Start position wrong\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-4512-5p\"][0] = \"15\"              # 3p is correct. \n",
    "mirge3_anno_dict[\"hsa-miR-4512-5p\"][1] =  66497012 \n",
    "mirge3_anno_dict[\"hsa-miR-4512-5p\"][2] =  66497034\n",
    "mirge3_anno_dict[\"hsa-miR-4512-5p\"][3] =  \"-\"\n",
    "\n",
    "# Removed because I don't trust it\n",
    "del mirge3_anno_dict[\"hsa-miR-3921-3p\"]                     # Indel based on assembly. Also extremely rare in our data. Perhaps miRNA sequence wrong?\n",
    "\n",
    "# Removed after validating against miRBase\n",
    "mirge3_anno_dict[\"hsa-miR-4801-5p\"][0] = \"4\"               # Incorrect position/chromosome. \n",
    "mirge3_anno_dict[\"hsa-miR-4801-5p\"][1] = 37241961\n",
    "mirge3_anno_dict[\"hsa-miR-4801-5p\"][2] = 37241984\n",
    "mirge3_anno_dict[\"hsa-miR-4801-5p\"][3] = \"-\"\n",
    "\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-663a-5p\"][0] = \"20\"              # 5p position wrong according to miRBase        \n",
    "mirge3_anno_dict[\"hsa-miR-663a-5p\"][1] =  26208243 \n",
    "mirge3_anno_dict[\"hsa-miR-663a-5p\"][2] =  26208264\n",
    "mirge3_anno_dict[\"hsa-miR-663a-5p\"][3] =  \"-\"\n",
    "mirge3_anno_dict[\"hsa-miR-663a-3p\"][0] = \"20\"              # 3p doesn't exist according to miRBase. But sequence exist on stem-loop.\n",
    "mirge3_anno_dict[\"hsa-miR-663a-3p\"][1] = 26208211\n",
    "mirge3_anno_dict[\"hsa-miR-663a-3p\"][2] = 26208228\n",
    "mirge3_anno_dict[\"hsa-miR-663a-3p\"][3] = \"-\"\n",
    "\n",
    "\n",
    "mirge3_anno_dict[\"hsa-miR-1256-5p\"][0] =  \"1\"             # 5p position wrong according to miRBase.        \n",
    "mirge3_anno_dict[\"hsa-miR-1256-5p\"][1] =  20988377        # 5p chromosome different than 3p.\n",
    "mirge3_anno_dict[\"hsa-miR-1256-5p\"][2] =  20988398\n",
    "mirge3_anno_dict[\"hsa-miR-1256-5p\"][3] =  \"-\"\n",
    "               \n",
    "\n",
    "mirge3_anno_dict['hsa-miR-4315-5p'][0] =  \"17\"             # Wrong position\n",
    "mirge3_anno_dict['hsa-miR-4315-5p'][1] = 45475408\n",
    "mirge3_anno_dict['hsa-miR-4315-5p'][2] = 45475425\n",
    "mirge3_anno_dict['hsa-miR-4315-5p'][3] = \"-\"\n",
    "mirge3_anno_dict['hsa-miR-4315-3p'][0] = '17'              # 3p doesn't exist according to miRBase. But sequence exist on stem-loop\n",
    "mirge3_anno_dict['hsa-miR-4315-3p'][1] = 45475377\n",
    "mirge3_anno_dict['hsa-miR-4315-3p'][2] = 45475395\n",
    "mirge3_anno_dict['hsa-miR-4315-3p'][3] = \"-\"\n",
    "\n",
    "\n",
    "mirge3_anno_dict['hsa-miR-3648-5p'][1] = 8208500           # Positions for 5p and 3p off. \n",
    "mirge3_anno_dict['hsa-miR-3648-5p'][2] = 8208520 \n",
    "mirge3_anno_dict['hsa-miR-3648-3p'][1] = 8208534\n",
    "mirge3_anno_dict['hsa-miR-3648-3p'][2] = 8208553\n",
    "\n",
    "\n",
    "mirge3_anno_dict['hsa-miR-4737-3p'][1] = 60043042          # Position for 3p slightly off (1000 bp)\n",
    "mirge3_anno_dict['hsa-miR-4737-3p'][2] = 60043062\n",
    "# 5p wasn't included for some reason. This is in miRBase database!\n",
    "mirge3_anno_dict['hsa-miR-4737-5p'] = (\"17\", 60043077, 60043096, \"-\", 'ATGCGAGGATGCTGACAGTG', \"GG\", \"CCTCAC\")\n",
    "\n",
    "mirge3_anno_dict['hsa-miR-1181-3p'][0] = '19'             # Position for 3p wrong. \n",
    "mirge3_anno_dict['hsa-miR-1181-3p'][1] = 10403468\n",
    "mirge3_anno_dict['hsa-miR-1181-3p'][2] = 10403490\n",
    "mirge3_anno_dict['hsa-miR-1181-3p'][3] = \"-\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "065397bf-92bb-417f-9890-2089a4703020",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of entries in mirge3 library (excluding SNPs): 2816\n"
     ]
    }
   ],
   "source": [
    "mirge3_anno_dict_no_SNPs = {} # key: name,   value: (chromosome, start, stop, strand, sequence, end5, end3)\n",
    "mirge3_anno_dict_SNPs = {}    # key: name,   value: [sequence_1, sequence_2, ...]\n",
    "\n",
    "for k, v in mirge3_anno_dict.items():\n",
    "    name = k.split(\".\")[0]\n",
    "    if (\"SNP\" in k) and (\"SNPC\" not in k):\n",
    "        if name not in mirge3_anno_dict_SNPs:\n",
    "            mirge3_anno_dict_SNPs[name] = [v[4]]\n",
    "        else:\n",
    "            mirge3_anno_dict_SNPs[name].append(v[4])\n",
    "    else:\n",
    "        mirge3_anno_dict_no_SNPs[name] = v\n",
    "print(f\"Number of entries in mirge3 library (excluding SNPs): {len(mirge3_anno_dict_no_SNPs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c124844f-9e91-42e0-bf4b-8f9919af0053",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of miRNA with SNPs: 238\n",
      "Number of SNPs: 335\n",
      "Number of SNPs that fall 3-nucleotides on the 5' or 3' end: 0\n"
     ]
    }
   ],
   "source": [
    "# Question: Where do SNPs occur in mirge3 library?\n",
    "positions_of_SNPs = {} # key: name    value: [[position, reference_allele, SNP], [posiiton, reference_allele, SNP], ...]\n",
    "in_isomiR_range = [] # isomiR define as SNPs 3-nucleotide deep on the 5' or 3' end. \n",
    "\n",
    "for k, SNPs in mirge3_anno_dict_SNPs.items():\n",
    "    cannoical_seq = mirge3_anno_dict_no_SNPs[k][4]\n",
    "    for SNP in SNPs:\n",
    "        position = [i for i in range(len(cannoical_seq)) if cannoical_seq[i] != SNP[i]]\n",
    "        if len(position) > 1:\n",
    "            print(f\"Error: Multiple SNP on same sequence!\")\n",
    "            \n",
    "        reference_allele =  cannoical_seq[position[0]]\n",
    "        SNP_allele = SNP[position[0]]\n",
    "        \n",
    "        if k in positions_of_SNPs:\n",
    "            positions_of_SNPs[k].append([position[0], reference_allele, SNP_allele])\n",
    "        else:\n",
    "            positions_of_SNPs[k] = [[position[0], reference_allele, SNP_allele]]\n",
    "        \n",
    "        if position in [0,1,2, len(SNP), len(SNP)-1, len(SNP)-2]:\n",
    "            in_isomiR_range.append(True)\n",
    "        else:\n",
    "            in_isomiR_range.append(False)\n",
    "\n",
    "print(f\"Number of miRNA with SNPs: {len(positions_of_SNPs)}\")\n",
    "print(f\"Number of SNPs: {len(in_isomiR_range)}\")\n",
    "print(f\"Number of SNPs that fall 3-nucleotides on the 5' or 3' end: {sum(in_isomiR_range)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51417825-d4d8-4967-baae-622e41e5eef1",
   "metadata": {},
   "source": [
    "### 2. Update annotations and check sequence against assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b8cc08f8-487b-43bd-8a10-5468b55a716b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_query(query, strand):\n",
    "    if strand == \"+\":\n",
    "        result = SP.check_output([str(samtools), \"faidx\", str(GRCh38), query, \"-n\", \"1000\"])\n",
    "    else:\n",
    "        result = SP.check_output([str(samtools), \"faidx\", str(GRCh38), query, \"-i\", \"-n\", \"1000\"])\n",
    "    result = result.decode('utf-8').split()\n",
    "    new_sequence = result[1].upper()\n",
    "    return new_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e201b689-2716-45f8-9b68-387a727e0156",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of miRNA with perfect matches for 5' and 3' ends: 2757\n",
      "Number of miRNA with imperfect matches for 5' and 3' ends: 59\n",
      "Number of miRNA with wrong positions: 232\n",
      "Number of miRNA with wrong positions fixed by adding +1 to start: 232\n",
      "Number of SNPs added: 0\n"
     ]
    }
   ],
   "source": [
    "mirge3_anno_dict_updated = {} # key: name,   value: (chromosome, new_start, stop, strand, prune_sequence, new_end5, new_end3)\n",
    "\n",
    "\n",
    "stat_wrong_position = 0 \n",
    "stat_wrong_position_fixed_by_StartPlusOne = 0\n",
    "stat_perfect_match = 0\n",
    "stat_inperfect_match = 0\n",
    "stat_number_of_SNPS = 0\n",
    "\n",
    "for miRNA, v in mirge3_anno_dict_no_SNPs.items():\n",
    "    perfect_match = True # For tracking 5' and 3' matches\n",
    "    chromosome, start, stop, strand, sequence, end5, end3 = v\n",
    "    chromosome_number = chromosome_to_GenBank[chromosome]\n",
    "    query = f\"{chromosome_number}:{start}-{stop}\"\n",
    "    \n",
    "    # Query Attempt #1\n",
    "    new_sequence = do_query(query, strand)\n",
    "    \n",
    "    # Query Attempt #2: Adds +1 to the start (Common error)\n",
    "    if sequence != new_sequence:\n",
    "        stat_wrong_position += 1\n",
    "        start = start + 1\n",
    "        query = f\"{chromosome_number}:{start}-{stop}\"\n",
    "        new_sequence = do_query(query, strand)\n",
    "        if sequence == new_sequence:\n",
    "            stat_wrong_position_fixed_by_StartPlusOne += 1\n",
    "    \n",
    "    # Final Check\n",
    "    if sequence != new_sequence:\n",
    "        print(f\"  {miRNA} strand:{strand} chr:{chromosome} start:{start}  stop:{stop}\")\n",
    "        print(f\"  Old sequence: {sequence}\")\n",
    "        print(f\"  New sequence: {new_sequence}\")\n",
    "        \n",
    "    # Get 5'-end and 3'-end\n",
    "    if strand == \"+\":\n",
    "        query = f\"{chromosome_number}:{start-2}-{start-1}\"\n",
    "        new_end5 = do_query(query, strand)\n",
    "        query = f\"{chromosome_number}:{stop+1}-{stop+6}\"\n",
    "        new_end3 = do_query(query, strand)\n",
    "    elif strand == \"-\":\n",
    "        query = f\"{chromosome_number}:{stop+1}-{stop+2}\"\n",
    "        new_end5 = do_query(query, strand)\n",
    "        query = f\"{chromosome_number}:{start-6}-{start-1}\"\n",
    "        new_end3 = do_query(query, strand)\n",
    "        \n",
    "    # Complete match?\n",
    "    if (end5 != new_end5) or (end3 != new_end3):\n",
    "        perfect_match = False\n",
    "        \n",
    "    if perfect_match:\n",
    "        stat_perfect_match += 1\n",
    "    else:\n",
    "        stat_inperfect_match += 1\n",
    "        \n",
    "    # Record\n",
    "    mirge3_anno_dict_updated[miRNA] = [chromosome, start, stop, strand, new_sequence, new_end5, new_end3]\n",
    "        \n",
    "print(f\"Number of miRNA with perfect matches for 5' and 3' ends: {stat_perfect_match}\")\n",
    "print(f\"Number of miRNA with imperfect matches for 5' and 3' ends: {stat_inperfect_match}\")\n",
    "print(f\"Number of miRNA with wrong positions: {stat_wrong_position}\")\n",
    "print(f\"Number of miRNA with wrong positions fixed by adding +1 to start: {stat_wrong_position_fixed_by_StartPlusOne}\")\n",
    "print(f\"Number of SNPs added: {stat_number_of_SNPS}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b92dbc-857b-4d57-9952-423a051eca0c",
   "metadata": {},
   "source": [
    "### 3. Validate annotations against miRBase.  \n",
    "- Note: A mature miRNA may align to multiple stem loops, which may come from multiple places on the genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "31291350-2cc8-45ad-aef5-f7a98e848c80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of entries in miRBase: 2652\n",
      "Number of miRBase miRNAs unique to miRge3: 318\n",
      "Number of miRBase miRNAs unique to miRBase: 154\n",
      "Number of miRBase miRNAs in common: 2498\n"
     ]
    }
   ],
   "source": [
    "miRBase_dict = {} # key: name,   value: [(chromosome,start, end), (chromosome,start, end), ...] \n",
    "pattern = \"Name=(.*)\"\n",
    "with open(miRBase, 'r') as f_in:\n",
    "    for line in f_in:\n",
    "        if line.startswith(\"#\"):\n",
    "                continue\n",
    "                \n",
    "        line_split = line.split()\n",
    "        chromosome = line_split[0][3::]\n",
    "        miRNA_type = line_split[2]\n",
    "        start = int(line_split[3])\n",
    "        stop = int(line_split[4])\n",
    "        if miRNA_type != \"miRNA\":\n",
    "            continue\n",
    "\n",
    "        name = re.search(pattern, line_split[8])\n",
    "        name = name.group(1).split(\";\")[0] # In case a ; follows the Name\n",
    "        \n",
    "        if name in miRBase_dict:\n",
    "            miRBase_dict[name].append((chromosome, start, stop))\n",
    "        else:\n",
    "            miRBase_dict[name] = [(chromosome, start, stop)]\n",
    "            \n",
    "print(f\"Number of entries in miRBase: {len(miRBase_dict)}\")\n",
    "\n",
    "unique_to_mirge3 = set(mirge3_anno_dict_updated.keys()).difference(miRBase_dict.keys())\n",
    "unique_to_miRBase = set(miRBase_dict.keys()).difference(mirge3_anno_dict_updated.keys())\n",
    "in_common = set(mirge3_anno_dict_updated.keys()).intersection(miRBase_dict.keys())\n",
    "print(f\"Number of miRBase miRNAs unique to miRge3: {len(unique_to_mirge3)}\")\n",
    "print(f\"Number of miRBase miRNAs unique to miRBase: {len(unique_to_miRBase)}\")\n",
    "print(f\"Number of miRBase miRNAs in common: {len(in_common)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "283b1f09-4132-4faf-a22e-43c4c6439227",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of mirge3 miRNA validated in miRBase: 2498\n"
     ]
    }
   ],
   "source": [
    "# Validated miRNA in common\n",
    "stat_number_validated = 0 \n",
    "for miRNA in in_common:\n",
    "    validated = False\n",
    "    \n",
    "    for miRBase_entry in miRBase_dict[miRNA]:\n",
    "        miRBase_chr, miRBase_start, miRBase_end = miRBase_entry\n",
    "        mirge3_chr, mirge3_start, mirge3_end = mirge3_anno_dict_updated[miRNA][0:3]\n",
    "        \n",
    "        if (miRBase_chr == mirge3_chr) and (miRBase_start == mirge3_start) and (miRBase_end == mirge3_end):\n",
    "            validated = True\n",
    "        \n",
    "    if validated:\n",
    "        stat_number_validated += 1\n",
    "    else:\n",
    "        print(f\"{miRNA} does not match!\")\n",
    "        print(f\"miRBase entry: {miRBase_dict[miRNA]}\")\n",
    "        print(f\"mirge3 entry: {mirge3_anno_dict_updated[miRNA]}\")\n",
    "        \n",
    "print(f\"Number of mirge3 miRNA validated in miRBase: {stat_number_validated}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "07553413-8b6b-4990-b5a7-34fbc6cf7cf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of mirge3 miRNA, which were thought to be unique, strict validated in miRBase: 154\n",
      "Number of mirge3 miRNA, which were thought to be unique, rough validated in miRBase: 153\n",
      "Number of mirge3 miRNA, which were thought to be unique, not validated in miRBase: 0\n"
     ]
    }
   ],
   "source": [
    "# Validated miRNA \"unique\" to miRge3\n",
    "# Note: Many of these miRNAs likely had 5p/3p information appended OR they include the both arms \n",
    "#  where miRBase didn't. \n",
    "\n",
    "stat_number_strict_validated = 0 # Exact position\n",
    "stat_number_rough_validated = 0  # Within 100 BP of each other\n",
    "stat_number_not_validated = 0\n",
    "\n",
    "for miRNA in unique_to_mirge3:\n",
    "    # Try removing arm information\n",
    "    new_name = \"-\".join(miRNA.split(\"-\")[:-1])\n",
    "    if new_name in miRBase_dict:\n",
    "        strict_validated = False\n",
    "        rough_validated = False\n",
    "        \n",
    "        for miRBase_entry in miRBase_dict[new_name]:\n",
    "            miRBase_chr, miRBase_start, miRBase_end = miRBase_entry\n",
    "            mirge3_chr, mirge3_start, mirge3_end = mirge3_anno_dict_updated[miRNA][0:3]\n",
    "            \n",
    "            if (miRBase_chr == mirge3_chr) and (miRBase_start == mirge3_start) and (miRBase_end == mirge3_end):\n",
    "                strict_validated = True\n",
    "            else: # rough validate check\n",
    "                pos_difference = abs(miRBase_start - mirge3_start)\n",
    "                if pos_difference <= 100:\n",
    "                    rough_validated = True\n",
    "\n",
    "        if strict_validated:\n",
    "            stat_number_strict_validated += 1\n",
    "        elif rough_validated:\n",
    "            stat_number_rough_validated += 1\n",
    "        else:\n",
    "            print(f\"{miRNA} does not match!\")\n",
    "            print(f\"miRBase entry: {miRBase_dict[new_name]}\")\n",
    "            print(f\"mirge3 entry: {mirge3_anno_dict_updated[miRNA]}\")\n",
    "            stat_number_not_validated += 1\n",
    "\n",
    "print(f\"Number of mirge3 miRNA, which were thought to be unique, strict validated in miRBase: {stat_number_strict_validated}\")\n",
    "print(f\"Number of mirge3 miRNA, which were thought to be unique, rough validated in miRBase: {stat_number_rough_validated}\")\n",
    "print(f\"Number of mirge3 miRNA, which were thought to be unique, not validated in miRBase: {stat_number_not_validated}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e751cbc-c108-45b9-94b7-a55f6310630b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3988d8-2cdf-4c68-99cb-82155e40a9e6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b246d2f-2a69-4e78-a63d-114105ffda73",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f552a45-6859-49d7-a400-354c29ce5c3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b9dda0-3166-49e9-a662-125cc4b17bab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fca5835-d519-40e8-8bd2-e08498e66232",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4fca8f8-906a-4b5e-8a23-dc849a254fc9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed8e6cb-213e-464e-a624-43cc8c2909e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2ec66ae-cab6-43eb-a2b6-ce1728d3ad81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82f7c432-7eb0-4ec0-8ba8-5e30ef658e85",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "385a0821-4c21-43e8-8f34-3ecf1564bcb2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c90664c-9559-4c6e-a7e5-2b055b2400ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cdbd9a14-a602-4874-8529-8f819042e03b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Recreate library file\n",
    "with open(output, 'w') as f_out:\n",
    "    for miRNA, v in mirge3_anno_dict_updated.items():\n",
    "        chromosome, start, stop, strand, sequence, end5, end3 = v\n",
    "        \n",
    "        if strand == \"+\":\n",
    "            start = start - 2\n",
    "            stop = stop + 6\n",
    "        else:\n",
    "            start = start - 6\n",
    "            stop = stop + 2\n",
    "            \n",
    "        new_sequence = end5+sequence+end3\n",
    "        new_sequence_length = len(new_sequence)\n",
    "        \n",
    "        line1 = f\">{miRNA} chr{chromosome} segs:1-{new_sequence_length} cds:{strand}:{start}-{stop}\\n\"\n",
    "        line2 = new_sequence+\"\\n\"\n",
    "        f_out.write(line1)\n",
    "        f_out.write(line2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c781951-840e-413a-bd3a-a733d9562a00",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
